File: app/models/events.py
"""SPEC §4.3's events block -- competitions, belt exams, seminars and trips (§5.8).

**An event never writes a billing table.** Plan W4: "Event fees call
`BillingService.create_charge(kind='event')`. The events lane never writes to a billing
table directly." So `event.fee_agorot` is the *price of the event* -- a setting -- while
what a family actually owes is a `charge`, reached through `event_registration.charge_id`.
There is deliberately no amount on the registration: a second copy of what is owed is a
second source of truth that will eventually disagree with the ledger.

**D9.2 -- there is no weight category anywhere in this module.** Artboard `7c` drew a
`משקל / קטגוריה` column and it is cut: §2.2 defers weight categories to v2 and they imply
`student` fields §4.3 does not carry. RSVP counts, parent consent and payment status --
the rest of `7c` -- all stand.
"""

from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
)
from sqlalchemy.dialects.postgresql import UUID as PGUUID
from sqlalchemy.orm import Mapped, mapped_column

from app.core.tenancy import TenantMixin
from app.models.base import Base, TimestampColumns, UUIDPrimaryKey

#: §4.3 -- `event  type(competition|belt_exam|seminar|joint_training|trip|other)`.
EVENT_TYPES = ("competition", "belt_exam", "seminar", "joint_training", "trip", "other")

#: §4.3 -- `event  status(draft|published|cancelled|completed)`. Nothing is visible to a
#: guardian while it is `draft`, which is what makes an event safe to build up over days.
EVENT_STATUSES = ("draft", "published", "cancelled", "completed")

#: §4.3 -- `event_target  target_type(studio|class|group|student)`. §5.8's targeting.
EVENT_TARGET_TYPES = ("studio", "class", "group", "student")

#: §4.3 -- `event_registration  rsvp(pending|yes|no)`. `pending` is a real state: nobody
#: has answered yet, which is different from having declined.
RSVP_STATES = ("pending", "yes", "no")

#: §4.3 -- `event_exam_result  result(pass|fail)`.
EXAM_RESULTS = ("pass", "fail")


class Event(UUIDPrimaryKey, TimestampColumns, TenantMixin, Base):
    """§4.3's `event`. Dashboard artboards `7a`, `7b`, `7c`.

    `location_id` **or** `location_text`: §5.8 allows an external venue that is not one of
    the studio's own locations, and a competition in another city is the common case. Both
    nullable, because a seminar at the club needs neither until it is scheduled.
    """

    __tablename__ = "event"
    __tenant_table_args__ = (
        CheckConstraint(
            "type IN ('competition', 'belt_exam', 'seminar', 'joint_training', 'trip', 'other')",
            name="event_type",
        ),
        CheckConstraint(
            "status IN ('draft', 'published', 'cancelled', 'completed')", name="event_status"
        ),
        CheckConstraint("ends_at > starts_at", name="event_time_range"),
        CheckConstraint("fee_agorot IS NULL OR fee_agorot >= 0", name="event_fee_non_negative"),
        # §5.8 -- consent text is required when consent is required, or the parent is
        # asked to agree to nothing.
        CheckConstraint(
            "requires_consent = false OR consent_text IS NOT NULL",
            name="event_consent_has_text",
        ),
        Index("ix_event_studio_id_starts_at", "studio_id", "starts_at"),
    )

    type: Mapped[str] = mapped_column(String(20), nullable=False)
    title: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    starts_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    ends_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    location_id: Mapped[uuid.UUID | None] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("location.id", ondelete="SET NULL")
    )
    #: An external venue. §5.8 -- a competition in another city is not a studio location.
    location_text: Mapped[str | None] = mapped_column(String(300))
    rsvp_deadline: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    #: The event's **price**, a setting. What a family owes is a `charge` -- see the
    #: module docstring.
    fee_agorot: Mapped[int | None] = mapped_column(Integer)
    requires_consent: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    consent_text: Mapped[str | None] = mapped_column(Text)
    status: Mapped[str] = mapped_column(String(10), nullable=False, default="draft")


class EventTarget(UUIDPrimaryKey, TimestampColumns, TenantMixin, Base):
    """§4.3 -- who the event is for. §5.8's targeting: the whole studio, a class, a group,
    or named students. Several rows compose, so "both beginner groups plus three seniors"
    is three rows rather than a query language.

    `target_id` carries no foreign key because the referent depends on `target_type`, and a
    polymorphic reference cannot have one. It is nullable for `target_type='studio'`, which
    names no particular row.
    """

    __tablename__ = "event_target"
    __tenant_table_args__ = (
        CheckConstraint(
            "target_type IN ('studio', 'class', 'group', 'student')", name="event_target_type"
        ),
        CheckConstraint(
            "target_type = 'studio' OR target_id IS NOT NULL", name="event_target_has_an_id"
        ),
        Index("uq_event_target", "event_id", "target_type", "target_id", unique=True),
    )

    event_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("event.id", ondelete="CASCADE"), nullable=False
    )
    target_type: Mapped[str] = mapped_column(String(10), nullable=False)
    target_id: Mapped[uuid.UUID | None] = mapped_column(PGUUID(as_uuid=True))


class EventRegistration(UUIDPrimaryKey, TimestampColumns, TenantMixin, Base):
    """§4.3 -- one student's participation in one event. Parent artboard `7d`.

    **No money column, deliberately.** `charge_id` points at the ledger's row; an
    `amount_agorot` here would be a second answer to what the family owes, and the two
    would diverge the first time a manager applied a discount. See the module docstring.

    `attended` is separate from `rsvp`: saying yes and turning up are different facts, and
    §5.8's post-event report is about the second one.
    """

    __tablename__ = "event_registration"
    __tenant_table_args__ = (
        CheckConstraint("rsvp IN ('pending', 'yes', 'no')", name="event_registration_rsvp"),
        Index("uq_event_registration", "event_id", "student_id", unique=True),
        Index("ix_event_registration_student_id", "student_id"),
    )

    event_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("event.id", ondelete="CASCADE"), nullable=False
    )
    student_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("student.id", ondelete="CASCADE"), nullable=False
    )
    rsvp: Mapped[str] = mapped_column(String(10), nullable=False, default="pending")
    responded_by_person_id: Mapped[uuid.UUID | None] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("person.id", ondelete="SET NULL")
    )
    responded_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    consent_signed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    #: The ledger's row. Created by `BillingService.create_charge(kind='event')` -- the
    #: events lane never writes a billing table itself.
    charge_id: Mapped[uuid.UUID | None] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("charge.id", ondelete="SET NULL")
    )
    attended: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)


class EventExamResult(UUIDPrimaryKey, TimestampColumns, TenantMixin, Base):
    """§4.3 -- `event_exam_result  event_id, student_id, belt_rank_id, result(pass|fail),
    examiner_person_id, note?`. Staff artboard `9d`, dashboard `4d`.

    A `fail` is recorded, not omitted. §5.9's eligibility view needs to know that a student
    was examined and did not pass -- an absent row would read as "never examined", which is
    a different conversation with a parent.
    """

    __tablename__ = "event_exam_result"
    __tenant_table_args__ = (
        CheckConstraint("result IN ('pass', 'fail')", name="event_exam_result_result"),
        Index("uq_event_exam_result", "event_id", "student_id", unique=True),
    )

    event_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("event.id", ondelete="CASCADE"), nullable=False
    )
    student_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("student.id", ondelete="CASCADE"), nullable=False
    )
    belt_rank_id: Mapped[uuid.UUID] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("belt_rank.id", ondelete="RESTRICT"), nullable=False
    )
    result: Mapped[str] = mapped_column(String(4), nullable=False)
    examiner_person_id: Mapped[uuid.UUID | None] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("person.id", ondelete="SET NULL")
    )
    note: Mapped[str | None] = mapped_column(Text)

File: web/packages/i18n/he/events.ts
import type { Bundle } from '../types'

/**
 * Owned by the EVENTS lane (M7). Hebrew is the reference locale — `en` and `ru` mirror
 * these keys and `web/scripts/i18n-parity.mjs events` fails on a gap in `en`.
 *
 * Artboards: dashboard `7a` אירועים ותחרויות, `7b` יצירת אירוע, `7c` עמוד אירוע (**D9.2**),
 * `4d` מבחן חגורה — זכאות וקידום, `6b` מבחני חגורה, `5b` מערכת חגורות, `5d` אשף · שלב 2;
 * staff `9d` מבחן חגורה, `9i` אירועים בצוות; parent `7d` הזמנה לאירוע, `12h` אירועים,
 * `12d` התקדמות חגורה.
 *
 * **Belt strings live here under `belt.*`, and that is deliberate.**
 * `web/packages/i18n/types.ts` lists exactly nine namespaces and `index.ts` is authored
 * once — a lane never edits it. §4.3 puts `belt_rank`/`student_belt` and `event` in the
 * same milestone (M7) and §5.9 makes a belt exam *an event with `type='belt_exam'`*, so
 * one namespace holding both is the shape the domain already has, not a workaround.
 *
 * Two rules the copy itself has to hold:
 *  - **D9.2 — there are no `משקל`/`קטגוריה` strings.** §2.2 defers weight categories to v2
 *    and they imply `student` fields §4.3 does not carry. Adding the key is how the cut
 *    quietly comes back.
 *  - **`draft` is invisible to guardians** (§4.3). `status.draftHint` says so on the screen,
 *    because an event built up over days is only safe if that is stated.
 */
export const events: Bundle = {
  // -- the events list (dashboard 7a, staff 9i, parent 12h) ----------------------
  'title': 'אירועים ותחרויות',
  'list.empty': 'אין אירועים מתוכננים',
  'list.upcoming': 'קרובים',
  'list.past': 'שהיו',
  'list.mine': 'האירועים שלי',
  'create': 'אירוע חדש',

  // -- an event's type and status ------------------------------------------------
  'type.competition': 'תחרות',
  'type.belt_exam': 'מבחן חגורה',
  'type.seminar': 'סמינר',
  'type.joint_training': 'אימון משותף',
  'type.trip': 'טיול',
  'type.other': 'אחר',
  'status.draft': 'טיוטה',
  'status.published': 'פורסם',
  'status.cancelled': 'בוטל',
  'status.completed': 'הסתיים',
  // §4.3 — nothing is visible to a guardian while the event is a draft.
  'status.draftHint': 'אירוע בטיוטה אינו מוצג להורים',
  'publish': 'פרסום האירוע',
  'published': 'האירוע פורסם',
  'cancel': 'ביטול האירוע',
  'cancelReason': 'סיבת הביטול',

  // -- creating an event (dashboard 7b) ------------------------------------------
  'form.title': 'יצירת אירוע',
  'form.name': 'שם האירוע',
  'form.description': 'תיאור',
  'form.type': 'סוג האירוע',
  'form.startsAt': 'מתחיל',
  'form.endsAt': 'מסתיים',
  'form.endBeforeStart': 'מועד הסיום חייב להיות אחרי מועד ההתחלה',
  'form.location': 'מיקום',
  'form.locationExternal': 'מקום חיצוני',
  // §5.8 — a competition in another city is not one of the studio's own locations.
  'form.locationExternalHint': 'לאולם או מקום שאינו אחד ממיקומי המועדון',
  'form.rsvpDeadline': 'הרשמה עד',
  'form.save': 'שמירה',
  'form.saveDraft': 'שמירה כטיוטה',

  // -- who it is for (§5.8's targeting) ------------------------------------------
  'target.title': 'קהל יעד',
  'target.studio': 'כל המועדון',
  'target.class': 'חוג',
  'target.group': 'קבוצה',
  'target.student': 'חניכים נבחרים',
  'target.add': 'הוספת קהל יעד',
  'target.empty': 'לא נבחר קהל יעד',
  'target.composeHint': 'אפשר לצרף כמה קהלים יחד',

  // -- the fee (§5.8 — the event's price is a setting; a charge is what a family owes)
  'fee.label': 'עלות',
  'fee.free': 'ללא עלות',
  'fee.perStudent': 'לחניך',
  'fee.chargeOnConfirm': 'אישור השתתפות יוצר חיוב להורה המשלם',

  // -- consent (§5.8) -------------------------------------------------------------
  'consent.required': 'דרוש אישור הורה',
  'consent.text': 'נוסח האישור',
  'consent.textRequired': 'אירוע הדורש אישור חייב לכלול נוסח',
  'consent.sign': 'אישור וחתימה',
  'consent.signed': 'האישור נחתם',
  'consent.pending': 'ממתין לאישור הורה',
  // §5.8 — the RSVP does not count as confirmed until the consent is signed.
  'consent.blocksConfirmation': 'ההשתתפות תיחשב מאושרת רק לאחר חתימת ההורה',

  // -- RSVP (parent 7d, dashboard 7c) ---------------------------------------------
  'rsvp.title': 'אישור השתתפות',
  'rsvp.yes': 'מגיע',
  'rsvp.no': 'לא מגיע',
  // §4.3 — `pending` is a real state: nobody answered, which is not the same as declining.
  'rsvp.pending': 'טרם ענו',
  'rsvp.answered': 'התשובה נשמרה',
  'rsvp.deadlinePassed': 'מועד ההרשמה חלף',
  'rsvp.change': 'שינוי התשובה',

  // -- the event page's counters (dashboard 7c ▲ D9.2 — no weight/category column) --
  'counts.registered': 'נרשמו',
  'counts.pending': 'טרם ענו',
  'counts.declined': 'לא מגיע',
  'counts.paid': 'שולם',
  'remindNonResponders': 'תזכורת למי שלא ענה',
  'reminderSent': 'התזכורת נשלחה',
  'roster.empty': 'אף חניך לא שויך לאירוע',

  'addToCalendar': 'הוסף ליומן',
  'attendance.take': 'סימון נוכחות באירוע',

  // -- belt exams (§5.9 — an exam is an event; staff 9d, dashboard 4d, 6b) ---------
  'exam.title': 'מבחן חגורה',
  'exam.plural': 'מבחני חגורה',
  'exam.candidates': 'מועמדים',
  'exam.nominate': 'שיבוץ מועמדים',
  'exam.eligibility': 'זכאות',
  'exam.eligibleHint': 'הזכאות מחושבת לפי הדרגה הנוכחית והוותק בה',
  'exam.notEligible': 'טרם זכאי',
  'exam.result.pass': 'עבר',
  'exam.result.fail': 'לא עבר',
  'exam.result.pending': 'טרם נבחן',
  'exam.note': 'הערת הבוחן',
  'exam.record': 'רישום תוצאות',
  'exam.recorded': 'התוצאות נרשמו',
  // §5.9 step 3 — pass writes the result, the belt row and the cache in one transaction.
  'exam.passPromotesHint': 'תוצאת ״עבר״ מעניקה את הדרגה הבאה ומעדכנת את כרטיס החניך',
  'exam.empty': 'לא נקבעו מבחני חגורה',

  // -- the belt system (dashboard 5b, wizard 5d) ----------------------------------
  'belt.title': 'מערכת חגורות',
  'belt.rank': 'דרגה',
  'belt.rankPlural': 'דרגות',
  'belt.add': 'דרגה חדשה',
  'belt.name': 'שם הדרגה',
  'belt.kyu': 'קיו',
  'belt.kyuOptional': 'לא כל מועדון משתמש בקיו',
  'belt.order': 'סדר',
  'belt.orderHint': 'הסדר קובע מהי הדרגה הבאה',
  'belt.color': 'צבע',
  'belt.secondaryColor': 'צבע משני',
  'belt.biColor': 'חגורה דו־צבעית',
  'belt.perClassHint': 'מערכת החגורות מוגדרת לכל חוג בנפרד',
  'belt.empty': 'לא הוגדרה מערכת חגורות',
  'belt.seedDefault': 'טעינת מערכת חגורות ברירת מחדל',

  // -- a student's belt (parent 12d, dashboard 4d) --------------------------------
  'belt.current': 'הדרגה הנוכחית',
  'belt.next': 'הדרגה הבאה',
  'belt.none': 'טרם הוענקה דרגה',
  'belt.progress': 'התקדמות חגורה',
  'belt.history': 'היסטוריית דרגות',
  'belt.awardedOn': 'הוענקה בתאריך',
  'belt.awardedBy': 'הוענקה על ידי',
  'belt.awardNote': 'הערה',
  'belt.award': 'הענקת דרגה',
  'belt.awarded': 'הדרגה הוענקה',
  // §5.9 — a promotion outside a formal exam is a real thing in a children's club.
  'belt.awardOutsideExam': 'הענקה ללא מבחן',
  'belt.groupPromote': 'קידום קבוצתי',
  'belt.groupPromoteHint': 'קידום כל המועמדים שעברו, בפעולה אחת',

  // -- 7a / 9i / 12h — list chrome the audits found missing ------------------------
  'list.loading': 'טוען…',
  'list.subtitle': 'אירועים חד-פעמיים — לא חלק מהלו״ז השבועי',
  'list.filterAll': 'הכל',
  'list.needsAttention': 'דורשים תשומת לב',
  // 7a finding 1 — `status.draftHint` says a draft is hidden from parents; this says why
  // the manager is still looking at it. 6b's draft copy is better than 7a's and this is it.
  'status.draftWhy': 'טיוטה — טרם הושלמה',

  // -- 7c / 9i — aggregates. The rsvp.* keys above are per-student and singular -------
  'counts.confirmed': 'אישרו',
  'counts.awaitingConsent': 'ללא אישור הורה',
  'counts.attended': 'הגיעו',

  // -- 7c — the participants table (D9.2 — six columns, none of them משקל) -----------
  'roster.title': 'רשימת משתתפים',
  'roster.columnConsent': 'אישור הורה חתום',
  'roster.columnPayment': 'תשלום',
  // The em dash on a cell that does not apply — a consent or a payment is meaningless
  // until someone has said yes. It needs a label, not a bare glyph.
  'roster.notApplicable': 'לא רלוונטי',
  'roster.sendConsentForm': 'שליחת טופס',

  // -- 7b findings 2 and 8 — a required field with no input, on a form that never errors
  'form.required': 'שדה חובה',
  'form.blank': 'אירוע חדש',
  'form.errorTitle': 'לא ניתן לשמור',
  'form.saved': 'האירוע נשמר',
  'form.edit': 'עריכת האירוע',

  // -- 7d / 12h finding 7 — the parent's screen speaks in the second person -----------
  'rsvp.awaitingYourAnswer': 'ממתין לתשובתכם',
  'rsvp.youConfirmed': 'אישרתם השתתפות',
  'rsvp.youDeclined': 'סימנתם שלא תגיעו',

  // -- 9d / 4d / 6b — the exam --------------------------------------------------------
  'exam.new': 'מבחן חגורה חדש',
  'exam.save': 'שמירת התוצאות',
  'exam.tenureAtRank': 'ותק בדרגה',
  'exam.readiness': 'מוכנות',
  // Deliberately impersonal. 4d finding 7: `מוכן`/`מוכנה` inflects per student and is the
  // first gendered STATUS value in the product. A neutral phrasing is the one thing this
  // lane can ship that is correct for every child.
  'exam.ready': 'עומד/ת בתנאים',
  'exam.confirmPromotion': 'אישור קידום',
  'exam.promoted': 'הדרגות הוענקו',

  // -- 5b / 5d — the belt system ------------------------------------------------------
  'form.cancel': 'ביטול',
  'belt.delete': 'מחיקת דרגה',
  'belt.edit': 'עריכת דרגה',
  'belt.save': 'שמירת דרגה',
  'belt.preview': 'תצוגה מקדימה',
  // 5b reorders by drag and there is no drag primitive and no shared drag utility, so the
  // rows move with buttons over `order_index` — the column that exists.
  'belt.moveUp': 'העלאה בסדר',
  'belt.moveDown': 'הורדה בסדר',
  // 5b finding 7 — the row already shows how many students hold the rank, so the refusal
  // has its reason on screen. `student_belt.belt_rank_id` is ON DELETE RESTRICT.
  'belt.deleteHeld': 'לא ניתן למחוק דרגה שהוענקה לחניכים',
  'belt.holders': 'חניכים בדרגה',
  'belt.noClassYet': 'מערכת החגורות תוגדר אחרי יצירת החוגים',
  'belt.presetTitle': 'איזו מערכת חגורות נהוגה אצלכם?',
  'belt.presetScratch': 'הגדרה ידנית',
  'belt.presetRankCount': 'דרגות בערכה',

  // -- 12d ------------------------------------------------------------------------------
  // 12d finding 7 — the artboard spells both ordinals as Hebrew words, which no
  // interpolation produces. Digits instead, rather than adding a Hebrew ordinal formatter
  // to `core`, which is not this lane's package.
  'belt.ordinalOfTotal': 'דרגה מתוך',
  'belt.progressCaption': 'הדרגות שהוענקו עד היום',
}
